torch/distributed/_composable/fsdp/_fsdp_state.py

```diff
@@ -19,13 +21,56 @@ class FSDPState(_State):
     def __init__(self):
         super().__init__()
         self._fsdp_param_group: Optional[FSDPParamGroup] = None
+        self._is_root: Optional[bool] = None
+        self._training_state: TrainingState = TrainingState.IDLE
+
+        # Attributes only used on the root state:
+        self._all_states: List[FSDPState] = []
 
     # Define a separate init since `__init__` is called in the contract
     def init(self, module: nn.Module, device: torch.device) -> None:
         _insert_module_state(module, self)
         self._module = module
         self._device = device
 
+    def _root_pre_forward(
```

Comment:
nit: could you add some comment about how root is determined and when we'll call `_root_pre_forward`?

Reply:
Related to the other comment: I refactored the root handling somewhat in https://github.com/pytorch/pytorch/pull/118004.

In summary:
- We define the root via lazy initialization. For now, this is only from the 1st forward, so we need the user to run forward through expected root on the 1st iteration. If there is some adversarial case to this, we can consider augmenting the logic.
- We define some logic that runs only once per forward and some logic that runs only once per backward. Instead of assuming that that logic always runs with the root defined from lazy initialization, we just make sure the first FSDP state to run that logic is the only one that runs it for that forward or that backward.
    - This gives greater flexibility. E.g., the user can run forward through a non-root without going through the root's forward and expect correct results. Without this, the non-root would miss the stream synchronization and could give wrong results.